compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -10287,6 +10356,98 @@ private STNode parseLetClause(boolean isRhsExpr) {
         return STNodeFactory.createLetClauseNode(letKeyword, letVarDeclarations);
     }
 
+    /**
+     * Parse order by clause.
+     * <p>
+     * <code>order-by-clause := order by expression (ascending | descending)? (expression (ascending | descending)?)*
+     * </code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseOrderByClause(boolean isRhsExpr) {
+        STToken nextToken = peek();
+        if (nextToken.kind != SyntaxKind.ORDER_KEYWORD) {
+            return STNodeFactory.createEmptyNode();
+        }
+        STNode orderKeyword = parseOrderKeyword();
+        STNode byKeyword = parseByKeyword();
+        STNode orderKeys = parseOrderKeyList(ParserRuleContext.ORDER_KEY, isRhsExpr);
+
+        byKeyword = cloneWithDiagnosticIfListEmpty(orderKeys, byKeyword,
+                DiagnosticErrorCode.ERROR_MISSING_EXPRESSION);
+
+        return STNodeFactory.createOrderByClauseNode(orderKeyword, byKeyword, orderKeys);
+    }
+
+    /**
+     * Parse order key.
+     * <p>
+     * <code>order-key-list := order-key [, order-key]*</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseOrderKeyList(ParserRuleContext context, boolean isRhsExpr) {
+        startContext(context);
+        List<STNode> orderKeys = new ArrayList<>();
+        STToken nextToken = peek();
+
+        // Make sure at least one order key is present
+        if (isEndOfOrderKeys(nextToken.kind)) {
+            endContext();
+            return STNodeFactory.createEmptyNodeList();
+        }
+
+        // Parse first order key, that has no leading comma
+        STNode orderKey = parseOrderKey(isRhsExpr);
+        orderKeys.add(orderKey);
+
+        // Parse the remaining order keys
+        nextToken = peek();
+        STNode leadingComma;
+        while (!isEndOfOrderKeys(nextToken.kind)) {
+            leadingComma = parseComma();
+            orderKeys.add(leadingComma);
+            orderKey = parseOrderKey(isRhsExpr);
+            orderKeys.add(orderKey);
+            nextToken = peek();
+        }
+
+        endContext();
+        return STNodeFactory.createNodeList(orderKeys);
+    }
+
+    private boolean isEndOfOrderKeys(SyntaxKind tokenKind) {
+        switch (tokenKind) {
+            case COMMA_TOKEN:
+                return false;
+            case IN_KEYWORD:
+                return true;
+            default:
+                return !isTypeStartingToken(tokenKind);
+        }
+    }
+
+    /**
+     * Parse order key.
+     * <p>
+     * <code>order-key := expression (ascending | descending)?</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseOrderKey(boolean isRhsExpr) {
+        STNode expression = parseExpression(OperatorPrecedence.QUERY, isRhsExpr, false);
+        STToken nextToken = peek();
+        if (nextToken.kind == SyntaxKind.ASCENDING_KEYWORD) {
```

Comment:
A switch-case would better suite for this.

We may have to go for error-recovery if the next token is neither `ascending`,`descending` nor a valid-token that is possible to come after the  order-key